### ISSUE TITLE
Registry

### ISSUE THREAD

**Issue by dominicbarnes** [duojs/duo#197]
So, I've been thinking about this a lot over the last couple of days. I wanted to start a discussion by sharing my thoughts on this, before I started writing any real code.

I think the best place to start is to have the registry _only_ responsible for searching. I don't think it's necessary to add tarballs or even deal with versioning at all.

With that being said, I think the registry would only need a few bits of data: (the rest lives in the source repository, so just a link should suffice here)
- name (user/repo)
- description
- keywords/tags
- stars/followers/etc

Now, onto the topic of _how_ data is added to the registry. After considering several alternatives, I think the best way to approach this is to use webhooks. We can make this extremely easy for duo users by having `duo-publish` use the [Github API](https://developer.github.com/webhooks/) to [add the webhook](https://developer.github.com/v3/repos/hooks/) for them. (then once the hook is in, no more work will be needed by the developer)

I'm thinking the "push" event is probably more than enough, although we can easily add more. We will need some sort of custom server ready to handle webhooks from various services. Github is obvious, but BitBucket support is in the works too, so we should probably have this server pluggable. (or at least easily extendable) It should probably support a "manual" API, allowing people to manually add their repo, without the continuous stream of webhooks.

I figure that each webhook call would trigger a "scrape" of the repository. (depending on what information comes in the payload of course) The `component.json` would be checked first, falling back to a `package.json`, and lastly to the repository meta itself. (all using the Github API I presume)

Lastly, I think that "keywords" can be used to group components in the search interface. I'm thinking duo can have a few that are special-cases that likely matches the structure of the wiki now. (eg: "ui-element", "utility", "async", etc) Beyond that level of structure, I think the rest of the fields are just searchable as plaintext.

As far as the technical implementation details, there are lots of possibilities. My first thoughts would be Heroku for the app, CouchDB for persistence and Elasticsearch for the actual searching/indexing. But depending on who would like to collaborate, how we would want to deploy, etc we can always work with other tools.

Anyways... sorry about the huge blob of text, as you can tell I've put a lot of thought into this lol
### tl;dr
- registry **only** be responsible for searching
- added to via Webhooks (and possibly expose a "manual" JSON API)
- uses component.json (or package.json)
- predefined keywords can be used by duo to group components

**Comment by johntron** [comment]
I created a super basic duo-search to save me a little time, but it just uses Github's search API to find JavaScript repos matching a keyword: https://github.com/johntron/duo-search. I'll happily transfer ownership.

**Comment by johntron** [comment]
@dominicbarnes I'd love to help setup the registry - have you already started?

**Comment by dominicbarnes** [comment]
@johntron I never really got anywhere with it, haven't taken the time. I would _love_ to see `duo-search` become a JS-api that can be consumed by multiple tools. (eg: a CLI, a web app, etc)

**Comment by johntron** [comment]
@dominicbarnes done: https://github.com/johntron/duo-search#usage-api
